From 56563c9083a01fc26d853edf4538ce70900d259c Mon Sep 17 00:00:00 2001
From: Mark Reynolds <mreynolds@redhat.com>
Date: Wed, 28 Jan 2026 12:02:54 -0500
Subject: [PATCH] Issue 7243 - UI - add support for hot certificates

Description:

In the "Add Server Certificate" modal add password options for pkcs#12
certificates. Also improved validation for certificate names

relates: https://github.com/389ds/389-ds-base/issues/7243

Reviewed by: jchapman & spichugi (Thanks!!)
---
 .../lib/security/certificateManagement.jsx    | 366 ++++++++++++---
 .../src/lib/security/securityModals.jsx       | 425 +++++++++++++++++-
 .../src/lib/security/securityTables.jsx       |  85 ++--
 src/cockpit/389-console/src/security.jsx      |  10 +
 src/lib389/lib389/cli_conf/security.py        |  23 +-
 5 files changed, 758 insertions(+), 151 deletions(-)

diff --git a/src/cockpit/389-console/src/lib/security/certificateManagement.jsx b/src/cockpit/389-console/src/lib/security/certificateManagement.jsx
index 475fcba27..18f98599b 100644
--- a/src/cockpit/389-console/src/lib/security/certificateManagement.jsx
+++ b/src/cockpit/389-console/src/lib/security/certificateManagement.jsx
@@ -19,6 +19,7 @@ import {
 import {
     EditCertModal,
     SecurityAddCertModal,
+    SecurityAddCACertModal,
     SecurityAddCSRModal,
     SecurityViewCSRModal,
     ExportCertModal,
@@ -37,6 +38,8 @@ export class CertificateManagement extends React.Component {
             CACerts: this.props.CACerts,
             ServerCSRs: this.props.ServerCSRs,
             ServerKeys: this.props.ServerKeys,
+            certNicknames: this.props.certNicknames,
+            CACertNicknames: this.props.CACertNicknames,
             tableKey: 0,
             showEditModal: false,
             showAddModal: false,
@@ -85,6 +88,11 @@ export class CertificateManagement extends React.Component {
             uploadFileName: "",
             uploadIsLoading: false,
             uploadIsRejected: false,
+            // PKCS#12 Password options
+            pkcs12PinMethod: "noPassword",
+            pkcs12PinFile: "",
+            pkcs12PinText: "",
+            forceCertAdd: false,
         };
 
         // File Upload functions
@@ -191,6 +199,11 @@ export class CertificateManagement extends React.Component {
         this.onRadioChange = this.onRadioChange.bind(this);
         this.validateCertText = this.validateCertText.bind(this);
         this.getCertFiles = this.getCertFiles.bind(this);
+        // PKCS#12 Password handlers
+        this.onPkcs12PinMethodChange = this.onPkcs12PinMethodChange.bind(this);
+        this.onPkcs12PinFileChange = this.onPkcs12PinFileChange.bind(this);
+        this.onPkcs12PinTextChange = this.onPkcs12PinTextChange.bind(this);
+        this.onForceCertAddChange = this.onForceCertAddChange.bind(this);
     }
 
     componentDidMount () {
@@ -211,12 +224,20 @@ export class CertificateManagement extends React.Component {
             certRadioUpload: true,
             isSelectCertOpen: false,
             modalSpinning: false,
+            pkcs12PinMethod: "noPassword",
+            pkcs12PinFile: "",
+            pkcs12PinText: "",
+            forceCertAdd: false,
         });
     }
 
     closeAddModal () {
         this.setState({
             showAddModal: false,
+            pkcs12PinMethod: "noPassword",
+            pkcs12PinFile: "",
+            pkcs12PinText: "",
+            forceCertAdd: false,
         });
     }
 
@@ -234,12 +255,20 @@ export class CertificateManagement extends React.Component {
             certRadioUpload: true,
             isSelectCertOpen: false,
             modalSpinning: false,
+            pkcs12PinMethod: "noPassword",
+            pkcs12PinFile: "",
+            pkcs12PinText: "",
+            forceCertAdd: false,
         });
     }
 
     closeAddCAModal () {
         this.setState({
             showAddCAModal: false,
+            pkcs12PinMethod: "noPassword",
+            pkcs12PinFile: "",
+            pkcs12PinText: "",
+            forceCertAdd: false,
         });
     }
 
@@ -263,6 +292,42 @@ export class CertificateManagement extends React.Component {
         });
     }
 
+    onPkcs12PinMethodChange(e, _) {
+        // Handle PKCS#12 password method selection
+        let pkcs12PinMethod = "";
+        if (e.target.id === "pkcs12PinRadioStdin") {
+            pkcs12PinMethod = "stdin";
+        } else if (e.target.id === "pkcs12PinRadioFile") {
+            pkcs12PinMethod = "file";
+        } else if (e.target.id === "noPasswordRadio") {
+            pkcs12PinMethod = "noPassword";
+        }
+        // Clear other password fields when method changes
+        this.setState({
+            pkcs12PinMethod,
+            pkcs12PinFile: "",
+            pkcs12PinText: "",
+        });
+    }
+
+    onPkcs12PinFileChange(value) {
+        this.setState({
+            pkcs12PinFile: value
+        });
+    }
+
+    onPkcs12PinTextChange(value) {
+        this.setState({
+            pkcs12PinText: value
+        });
+    }
+
+    onForceCertAddChange(checked) {
+        this.setState({
+            forceCertAdd: checked
+        });
+    }
+
     showAddCSRModal () {
         this.setState({
             showAddCSRModal: true,
@@ -468,43 +533,112 @@ export class CertificateManagement extends React.Component {
                             "dsconf", "-j", "ldapi://%2fvar%2frun%2fslapd-" + this.props.serverId + ".socket",
                             "security", certType, "add", "--name=" + this.state.certName, "--file=" + certFile
                         ];
+
+                        // Add PKCS#12 password options
+                        if (this.state.pkcs12PinMethod === "file" && this.state.pkcs12PinFile !== "") {
+                            cmd.push("--pkcs12-pin-path=" + this.state.pkcs12PinFile);
+                        } else if (this.state.pkcs12PinMethod === "stdin") {
+                            cmd.push("--pkcs12-pin-stdin");
+                        }
+
+                        // Add force flag
+                        if (this.state.forceCertAdd) {
+                            cmd.push("--do-it");
+                        }
+
                         log_cmd("addCert", "Adding cert (tmp): ", cmd);
-                        cockpit
-                                .spawn(cmd, { superuser: true, err: "message" })
-                                .done(() => {
-                                    this.deleteTmpCert(certFile);
-                                    this.reloadCACerts();
-                                    this.setState({
-                                        showAddModal: false,
-                                        modalSpinning: false,
-                                        loading: false,
+
+                        // Handle stdin method with PTY spawn
+                        if (this.state.pkcs12PinMethod === "stdin") {
+                            let buffer = "";
+                            const proc = cockpit.spawn(cmd, { pty: true, environ: ["LC_ALL=C"], superuser: true, err: "message" });
+                            proc
+                                    .done(() => {
+                                        this.deleteTmpCert(certFile);
+                                        this.reloadCACerts();
+                                        this.setState({
+                                            showAddModal: false,
+                                            modalSpinning: false,
+                                            loading: false,
+                                        });
+                                        this.reloadOrphanKeys();
+                                        this.props.addNotification(
+                                            "success",
+                                            _("Successfully added certificate")
+                                        );
+                                        this.closeAddCAModal();
+                                        this.closeAddModal();
+                                    })
+                                    .fail(err => {
+                                        let msg = _("Unknown error");
+                                        try {
+                                            const errMsg = buffer ? JSON.parse(buffer) : JSON.parse(err);
+                                            msg = errMsg.desc || errMsg.message || buffer || err;
+                                            if ('info' in errMsg) {
+                                                msg = errMsg.desc + " - " + errMsg.info;
+                                            }
+                                        } catch (e) {
+                                            // If parsing fails, use buffer or error directly
+                                            msg = buffer || err || _("Unknown error");
+                                        }
+                                        this.deleteTmpCert(certFile);
+                                        this.closeAddCAModal();
+                                        this.closeAddModal();
+                                        this.setState({
+                                            modalSpinning: false,
+                                            loading: false,
+                                        });
+                                        this.props.addNotification(
+                                            "error",
+                                            cockpit.format(_("Error adding certificate - $0"), msg)
+                                        );
+                                    })
+                                    .stream(data => {
+                                        buffer += data;
+                                        const lines = buffer.split("\n");
+                                        const last_line = lines[lines.length - 1].toLowerCase();
+                                        if (last_line.includes("password") || last_line.includes("pin")) {
+                                            proc.input(this.state.pkcs12PinText + "\n", true);
+                                        }
                                     });
-                                    this.reloadOrphanKeys();
-                                    this.props.addNotification(
-                                        "success",
-                                        _("Successfully added certificate")
-                                    );
-                                    this.closeAddCAModal();
-                                    this.closeAddModal();
-                                })
-                                .fail(err => {
-                                    const errMsg = JSON.parse(err);
-                                    let msg = errMsg.desc;
-                                    if ('info' in errMsg) {
-                                        msg = errMsg.desc + " - " + errMsg.info;
-                                    }
-                                    this.deleteTmpCert(certFile);
-                                    this.closeAddCAModal();
-                                    this.closeAddModal();
-                                    this.setState({
-                                        modalSpinning: false,
-                                        loading: false,
+                        } else {
+                            cockpit
+                                    .spawn(cmd, { superuser: true, err: "message" })
+                                    .done(() => {
+                                        this.deleteTmpCert(certFile);
+                                        this.reloadCACerts();
+                                        this.setState({
+                                            showAddModal: false,
+                                            modalSpinning: false,
+                                            loading: false,
+                                        });
+                                        this.reloadOrphanKeys();
+                                        this.props.addNotification(
+                                            "success",
+                                            _("Successfully added certificate")
+                                        );
+                                        this.closeAddCAModal();
+                                        this.closeAddModal();
+                                    })
+                                    .fail(err => {
+                                        const errMsg = JSON.parse(err);
+                                        let msg = errMsg.desc;
+                                        if ('info' in errMsg) {
+                                            msg = errMsg.desc + " - " + errMsg.info;
+                                        }
+                                        this.deleteTmpCert(certFile);
+                                        this.closeAddCAModal();
+                                        this.closeAddModal();
+                                        this.setState({
+                                            modalSpinning: false,
+                                            loading: false,
+                                        });
+                                        this.props.addNotification(
+                                            "error",
+                                            cockpit.format(_("Error adding certificate - $0"), msg)
+                                        );
                                     });
-                                    this.props.addNotification(
-                                        "error",
-                                        cockpit.format(_("Error adding certificate - $0"), msg)
-                                    );
-                                });
+                        }
                     })
                     .fail(err => {
                         this.setState({
@@ -529,42 +663,110 @@ export class CertificateManagement extends React.Component {
                 // certRadioSelect
                 cmd.push("--file=" + this.props.certDir + "/" + this.state.selectCertName);
             }
+
+            // Add PKCS#12 password options
+            if (this.state.pkcs12PinMethod === "file" && this.state.pkcs12PinFile !== "") {
+                cmd.push("--pkcs12-pin-path=" + this.state.pkcs12PinFile);
+            } else if (this.state.pkcs12PinMethod === "stdin") {
+                cmd.push("--pkcs12-pin-stdin");
+            }
+
+            // Add force flag
+            if (this.state.forceCertAdd) {
+                cmd.push("--do-it");
+            }
+
             log_cmd("addCert", "Adding cert: ", cmd);
-            cockpit
-                    .spawn(cmd, { superuser: true, err: "message" })
-                    .done(() => {
-                        this.reloadCACerts();
-                        this.closeAddCAModal();
-                        this.closeAddModal();
-                        this.setState({
-                            showAddModal: false,
-                            certFile: '',
-                            certName: '',
-                            modalSpinning: false
+
+            // Handle stdin method with PTY spawn
+            if (this.state.pkcs12PinMethod === "stdin") {
+                let buffer = "";
+                const proc = cockpit.spawn(cmd, { pty: true, environ: ["LC_ALL=C"], superuser: true, err: "message" });
+                proc
+                        .done(() => {
+                            this.reloadCACerts();
+                            this.closeAddCAModal();
+                            this.closeAddModal();
+                            this.setState({
+                                showAddModal: false,
+                                certFile: '',
+                                certName: '',
+                                modalSpinning: false
+                            });
+                            this.reloadOrphanKeys();
+                            this.props.addNotification(
+                                "success",
+                                _("Successfully added certificate")
+                            );
+                        })
+                        .fail(err => {
+                            let msg = _("Unknown error");
+                            try {
+                                const errMsg = buffer ? JSON.parse(buffer) : JSON.parse(err);
+                                msg = errMsg.desc || errMsg.message || buffer || err;
+                                if ('info' in errMsg) {
+                                    msg = errMsg.desc + " - " + errMsg.info;
+                                }
+                            } catch (e) {
+                                // If parsing fails, use buffer or error directly
+                                msg = buffer || err || _("Unknown error");
+                            }
+                            this.closeAddCAModal();
+                            this.closeAddModal();
+                            this.setState({
+                                modalSpinning: false,
+                                loading: false,
+                            });
+                            this.props.addNotification(
+                                "error",
+                                cockpit.format(_("Error adding certificate - $0"), msg)
+                            );
+                        })
+                        .stream(data => {
+                            buffer += data;
+                            const lines = buffer.split("\n");
+                            const last_line = lines[lines.length - 1].toLowerCase();
+                            if (last_line.includes("password") || last_line.includes("pin")) {
+                                proc.input(this.state.pkcs12PinText + "\n", true);
+                            }
                         });
-                        this.reloadOrphanKeys();
-                        this.props.addNotification(
-                            "success",
-                            _("Successfully added certificate")
-                        );
-                    })
-                    .fail(err => {
-                        const errMsg = JSON.parse(err);
-                        let msg = errMsg.desc;
-                        if ('info' in errMsg) {
-                            msg = errMsg.desc + " - " + errMsg.info;
-                        }
-                        this.closeAddCAModal();
-                        this.closeAddModal();
-                        this.setState({
-                            modalSpinning: false,
-                            loading: false,
+            } else {
+                cockpit
+                        .spawn(cmd, { superuser: true, err: "message" })
+                        .done(() => {
+                            this.reloadCACerts();
+                            this.closeAddCAModal();
+                            this.closeAddModal();
+                            this.setState({
+                                showAddModal: false,
+                                certFile: '',
+                                certName: '',
+                                modalSpinning: false
+                            });
+                            this.reloadOrphanKeys();
+                            this.props.addNotification(
+                                "success",
+                                _("Successfully added certificate")
+                            );
+                        })
+                        .fail(err => {
+                            const errMsg = JSON.parse(err);
+                            let msg = errMsg.desc;
+                            if ('info' in errMsg) {
+                                msg = errMsg.desc + " - " + errMsg.info;
+                            }
+                            this.closeAddCAModal();
+                            this.closeAddModal();
+                            this.setState({
+                                modalSpinning: false,
+                                loading: false,
+                            });
+                            this.props.addNotification(
+                                "error",
+                                cockpit.format(_("Error adding certificate - $0"), msg)
+                            );
                         });
-                        this.props.addNotification(
-                            "error",
-                            cockpit.format(_("Error adding certificate - $0"), msg)
-                        );
-                    });
+            }
         }
     }
 
@@ -1088,7 +1290,8 @@ export class CertificateManagement extends React.Component {
                         ServerCerts: certs,
                         loading: false,
                         tableKey: key,
-                        showConfirmCAChange: false
+                        showConfirmCAChange: false,
+                        certNicknames: certNames,
                     }, this.getCertFiles);
                 })
                 .fail(err => {
@@ -1180,9 +1383,14 @@ export class CertificateManagement extends React.Component {
                 .spawn(cmd, { superuser: true, err: "message" })
                 .done(content => {
                     const certs = JSON.parse(content);
+                    const certNames = [];
+                    for (const cert of certs) {
+                        certNames.push(cert.attrs.nickname);
+                    }
                     this.setState({
                         CACerts: certs,
-                        loading: false
+                        loading: false,
+                        CACertNicknames: certNames,
                     }, this.reloadCerts);
                 })
                 .fail(err => {
@@ -1311,6 +1519,7 @@ export class CertificateManagement extends React.Component {
                     spinning={this.state.modalSpinning}
                 />
                 <SecurityAddCertModal
+                    key={"addCert-" + this.state.showAddModal}
                     showModal={this.state.showAddModal}
                     closeHandler={this.closeAddModal}
                     handleChange={this.onChange}
@@ -1319,6 +1528,8 @@ export class CertificateManagement extends React.Component {
                     certFile={this.state.certFile}
                     certName={this.state.certName}
                     certNames={this.state.availCertNames}
+                    certNicknames={this.state.certNicknames}
+                    CACertNicknames={this.state.CACertNicknames}
                     selectCertName={this.state.selectCertName}
                     isSelectCertOpen={this.state.isSelectCertOpen}
                     handleCertSelect={this.onCertSelect}
@@ -1336,11 +1547,18 @@ export class CertificateManagement extends React.Component {
                     handleFileReadStarted={this.onFileReadStarted}
                     handleFileReadFinished={this.onFileReadFinished}
                     handleClear={this.onClear}
-                    handleFileRejected={this.state.uploadIsRejected}
+                    handleFileRejected={this.handleFileRejected}
+                    pkcs12PinMethod={this.state.pkcs12PinMethod}
+                    pkcs12PinFile={this.state.pkcs12PinFile}
+                    pkcs12PinText={this.state.pkcs12PinText}
+                    forceCertAdd={this.state.forceCertAdd}
+                    handlePkcs12PinMethodChange={this.onPkcs12PinMethodChange}
+                    handlePkcs12PinFileChange={this.onPkcs12PinFileChange}
+                    handlePkcs12PinTextChange={this.onPkcs12PinTextChange}
+                    handleForceCertAddChange={this.onForceCertAddChange}
                 />
-                <SecurityAddCertModal
+                <SecurityAddCACertModal
                     showModal={this.state.showAddCAModal}
-                    isCACert
                     closeHandler={this.closeAddCAModal}
                     handleChange={this.onChange}
                     saveHandler={this.addCert}
@@ -1348,6 +1566,8 @@ export class CertificateManagement extends React.Component {
                     certFile={this.state.certFile}
                     certName={this.state.certName}
                     certNames={this.state.availCertNames}
+                    certNicknames={this.state.certNicknames}
+                    CACertNicknames={this.state.CACertNicknames}
                     selectCertName={this.state.selectCertName}
                     isSelectCertOpen={this.state.isSelectCertOpen}
                     handleCertSelect={this.onCertSelect}
@@ -1365,7 +1585,7 @@ export class CertificateManagement extends React.Component {
                     handleFileReadStarted={this.onFileReadStarted}
                     handleFileReadFinished={this.onFileReadFinished}
                     handleClear={this.onClear}
-                    handleFileRejected={this.state.uploadIsRejected}
+                    handleFileRejected={this.handleFileRejected}
                 />
                 <SecurityAddCSRModal
                     showModal={this.state.showAddCSRModal}
diff --git a/src/cockpit/389-console/src/lib/security/securityModals.jsx b/src/cockpit/389-console/src/lib/security/securityModals.jsx
index a42aa50d6..4d6631fd2 100644
--- a/src/cockpit/389-console/src/lib/security/securityModals.jsx
+++ b/src/cockpit/389-console/src/lib/security/securityModals.jsx
@@ -6,6 +6,7 @@ import {
 	ClipboardCopy,
 	ClipboardCopyVariant,
 	Divider,
+	ExpandableSection,
 	FileUpload,
 	Form,
 	FormSelect,
@@ -22,12 +23,13 @@ import {
 	TextVariants,
 	TextInput,
 	Tooltip,
+	TooltipPosition,
 	ValidatedOptions
 } from '@patternfly/react-core';
 import TypeaheadSelect from "../../dsBasicComponents.jsx";
 import { OutlinedQuestionCircleIcon } from '@patternfly/react-icons/dist/js/icons/outlined-question-circle-icon';
 import PropTypes from "prop-types";
-import { bad_file_name, validHostname } from "../tools.jsx";
+import { bad_file_name, validHostname, file_is_path } from "../tools.jsx";
 
 const _ = cockpit.gettext;
 
@@ -124,6 +126,20 @@ export class ExportCertModal extends React.Component {
 }
 
 export class SecurityAddCertModal extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            isPasswordSectionExpanded: false
+        };
+        this.handlePasswordSectionToggle = this.handlePasswordSectionToggle.bind(this);
+    }
+
+    handlePasswordSectionToggle(event, isExpanded) {
+        this.setState({
+            isPasswordSectionExpanded: isExpanded
+        });
+    }
+
     render() {
         const {
             showModal,
@@ -139,6 +155,8 @@ export class SecurityAddCertModal extends React.Component {
             handleRadioChange,
             badCertText,
             certNames,
+            certNicknames,
+            CACertNicknames,
             // Select server cert
             handleCertSelect,
             selectCertName,
@@ -153,7 +171,15 @@ export class SecurityAddCertModal extends React.Component {
             handleFileReadFinished,
             handleClear,
             handleFileRejected,
-            isCACert,
+            // PKCS#12 Password options
+            pkcs12PinMethod,
+            pkcs12PinFile,
+            pkcs12PinText,
+            forceCertAdd,
+            handlePkcs12PinMethodChange,
+            handlePkcs12PinFileChange,
+            handlePkcs12PinTextChange,
+            handleForceCertAddChange,
         } = this.props;
 
         let saveBtnName = _("Add Certificate");
@@ -179,11 +205,358 @@ export class SecurityAddCertModal extends React.Component {
 
         let title = _("Add Server Certificate");
         let desc = _("Add a Server Certificate to the security database.");
-        if (isCACert) {
-            title = _("Add Certificate Authority");
-            desc = _("Add a CA Certificate to the security database.");
+
+        let selectValidated = ValidatedOptions.default;
+        if (certRadioSelect && certNames.length === 0) {
+            selectValidated = ValidatedOptions.error;
+        }
+
+        return (
+            <Modal
+                variant={ModalVariant.medium}
+                title={title}
+                aria-labelledby="ds-modal"
+                isOpen={showModal}
+                onClose={closeHandler}
+                actions={[
+                    <Button
+                        key="confirm"
+                        variant="primary"
+                        onClick={() => {
+                            saveHandler(false);
+                        }}
+                        isLoading={spinning}
+                        spinnerAriaValueText={spinning ? _("Saving") : undefined}
+                        {...extraPrimaryProps}
+                        isDisabled={
+                            certName === "" || (certRadioFile && certFile === "") ||
+                            (certRadioUpload && (uploadValue === "" || badCertText)) ||
+                            (certRadioSelect && certNames.length === 0) ||
+                            (pkcs12PinMethod === "file" && (pkcs12PinFile === "" || !file_is_path(pkcs12PinFile))) ||
+                            (pkcs12PinMethod === "stdin" && pkcs12PinText === "") ||
+                            certNicknames.includes(certName) ||
+                            CACertNicknames.includes(certName)
+                        }
+                    >
+                        {saveBtnName}
+                    </Button>,
+                    <Button key="cancel" variant="link" onClick={closeHandler}>
+                        {_("Cancel")}
+                    </Button>
+                ]}
+            >
+                <Form isHorizontal autoComplete="off">
+                    <TextContent>
+                        <Text component={TextVariants.h4}>
+                            {desc}
+                        </Text>
+                    </TextContent>
+                    <Grid
+                        className="ds-margin-top-sm"
+                        title={_("Enter name/nickname of the certificate")}
+                    >
+                        <GridItem className="ds-label" span={3}>
+                            {_("Certificate Nickname")}
+                        </GridItem>
+                        <GridItem span={9}>
+                            <TextInput
+                                type="text"
+                                id="certName"
+                                aria-describedby="horizontal-form-name-helper"
+                                name="certName"
+                                onChange={(e, str) => {
+                                    handleChange(e);
+                                }}
+                                validated={
+                                    certName === "" ||
+                                    certNicknames.includes(certName) ||
+                                    CACertNicknames.includes(certName) ? ValidatedOptions.error : ValidatedOptions.default}
+                            />
+                            {(certNicknames.includes(certName) || CACertNicknames.includes(certName)) && (
+                                <HelperText>
+                                    <HelperTextItem variant="error">
+                                        {_("Please use a unique certificate nickname")}
+                                    </HelperTextItem>
+                                </HelperText>
+                            )}
+                        </GridItem>
+                    </Grid>
+                    <ExpandableSection
+                        toggleText={this.state.isPasswordSectionExpanded ? _("Hide PKCS#12 Password Options") : _("Show PKCS#12 Password Options")}
+                        onToggle={this.handlePasswordSectionToggle}
+                        isExpanded={this.state.isPasswordSectionExpanded}
+                    >
+                        <div className="ds-indent-lg">
+                            <TextContent className="ds-margin-top">
+                                <Text component={TextVariants.p}>
+                                    {_("These password settings are only used for PKCS#12 password protected certificates. These settings are ignored for other certificate types.")}
+                                </Text>
+                            </TextContent>
+                            <Grid className="ds-margin-top-lg">
+                                <GridItem span={12}>
+                                    <Radio
+                                        id="noPasswordRadio"
+                                        label={_("Non-PKCS#12 Certificate")}
+                                        name="pkcs12PinMethod"
+                                        isChecked={pkcs12PinMethod === "noPassword"}
+                                        onChange={handlePkcs12PinMethodChange}
+                                    />
+                                    <div
+                                        className="ds-margin-top"
+                                        title={_("Password will be read from stdin when prompted by the command")}
+                                    >
+                                        <Radio
+                                            id="pkcs12PinRadioStdin"
+                                            label={_("Password")}
+                                            name="pkcs12PinMethod"
+                                            isChecked={pkcs12PinMethod === "stdin"}
+                                            onChange={handlePkcs12PinMethodChange}
+                                        />
+                                    </div>
+                                    <div className="ds-margin-top ds-radio-indent">
+                                        <TextInput
+                                            type="password"
+                                            id="pkcs12PinTextStdin"
+                                            aria-describedby="pkcs12-pin-stdin-helper"
+                                            name="pkcs12PinText"
+                                            value={pkcs12PinText}
+                                            onChange={(e, value) => {
+                                                handlePkcs12PinTextChange(value);
+                                            }}
+                                            placeholder={_("Enter password to send via stdin")}
+                                            validated={pkcs12PinMethod === "stdin" && pkcs12PinText === "" ? ValidatedOptions.error : ValidatedOptions.default}
+                                            isDisabled={pkcs12PinMethod !== "stdin"}
+                                        />
+                                    </div>
+                                    <div
+                                        className="ds-margin-top"
+                                        title={_("Read password from a file on the server")}
+                                    >
+                                        <Radio
+                                            id="pkcs12PinRadioFile"
+                                            className="ds-margin-top-lg"
+                                            label={_("Read password from file")}
+                                            name="pkcs12PinMethod"
+                                            isChecked={pkcs12PinMethod === "file"}
+                                            onChange={handlePkcs12PinMethodChange}
+                                        />
+                                    </div>
+                                    <div className="ds-margin-top ds-radio-indent">
+                                        <TextInput
+                                            type="text"
+                                            id="pkcs12PinFile"
+                                            aria-describedby="pkcs12-pin-file-helper"
+                                            name="pkcs12PinFile"
+                                            value={pkcs12PinFile}
+                                            onChange={(e, value) => {
+                                                handlePkcs12PinFileChange(value);
+                                            }}
+                                            placeholder={_("Enter full path to password file")}
+                                            isDisabled={pkcs12PinMethod !== "file"}
+                                            validated={
+                                                pkcs12PinMethod === "file" &&
+                                                (pkcs12PinFile === "" || !file_is_path(pkcs12PinFile))
+                                                    ? ValidatedOptions.error
+                                                    : ValidatedOptions.default
+                                            }
+                                        />
+                                        {pkcs12PinMethod === "file" && pkcs12PinFile !== "" && !file_is_path(pkcs12PinFile) && (
+                                            <HelperText>
+                                                <HelperTextItem variant="error">
+                                                    {_("Please enter a valid file path (must start with '/' and not end with '/')")}
+                                                </HelperTextItem>
+                                            </HelperText>
+                                        )}
+                                    </div>
+                                    <div className="ds-margin-top-xxlg">
+                                        <Tooltip
+                                            position={TooltipPosition.bottomStart}
+                                            content={
+                                                <div>
+                                                    {_("Force certificate addition without validation. This bypasses certificate chain validation checks.")}
+                                                </div>
+                                            }
+                                        >
+                                            <Checkbox
+                                                id="forceCertAdd"
+                                                className="ds-margin-top-lg"
+                                                label={_("Skip certificate verification")}
+                                                isChecked={forceCertAdd}
+                                                onChange={(e, checked) => {
+                                                    handleForceCertAddChange(checked);
+                                                }}
+                                            />
+                                        </Tooltip>
+                                    </div>
+                                </GridItem>
+                            </Grid>
+                        </div>
+                    </ExpandableSection>
+                    <Grid>
+                        <GridItem span={12}>
+                            <div title={_("Upload the contents of a PEM file from the client's system.")}>
+                                <Radio
+                                    id="certRadioUpload"
+                                    label={certTextLabel}
+                                    name="certChoice"
+                                    onChange={handleRadioChange}
+                                    isChecked={certRadioUpload}
+                                />
+                            </div>
+                            <div className={certRadioUpload ? "ds-margin-top ds-radio-indent" : "ds-margin-top ds-radio-indent ds-disabled"}>
+                                <FileUpload
+                                    id="uploadPEMFile"
+                                    type="text"
+                                    value={uploadValue}
+                                    filename={uploadFileName}
+                                    filenamePlaceholder={_("Drag and drop a file, or upload one")}
+                                    onFileInputChange={handleFileInputChange}
+                                    onDataChange={handleTextOrDataChange}
+                                    onTextChange={handleTextOrDataChange}
+                                    onReadStarted={handleFileReadStarted}
+                                    onReadFinished={handleFileReadFinished}
+                                    onClearClick={handleClear}
+                                    isLoading={uploadIsLoading}
+                                    dropzoneProps={{
+                                        accept: '.pem',
+                                        onDropRejected: handleFileRejected
+                                    }}
+                                    validated={
+                                        uploadIsRejected ||
+                                        (certRadioUpload && uploadValue === "") ||
+                                        (certRadioUpload && badCertText)
+                                            ? 'error'
+                                            : 'default'
+                                    }
+                                    browseButtonText={_("Upload PEM File")}
+                                />
+                            </div>
+                            <div title={_("Choose a certificate from the server's certificate directory")}>
+                                <Radio
+                                    id="certRadioSelect"
+                                    className="ds-margin-top-lg"
+                                    label={_("Choose Certificate From Server")}
+                                    name="certChoice"
+                                    isChecked={certRadioSelect}
+                                    onChange={handleRadioChange}
+                                />
+                            </div>
+                            <div className={certRadioSelect ? "ds-margin-top ds-radio-indent" : "ds-margin-top ds-radio-indent ds-disabled"}>
+                                <FormSelect
+                                    value={selectCertName}
+                                    id="selectCertName"
+                                    onChange={(e, str) => {
+                                        handleCertSelect(str);
+                                    }}
+                                    aria-label="FormSelect Input"
+                                    className="ds-cert-select"
+                                    validated={selectValidated}
+                                >
+                                    {certNames.length === 0 &&
+                                        <FormSelectOption
+                                            key="none"
+                                            value=""
+                                            label={_("No certificates present")}
+                                            isDisabled
+                                            isPlaceholder
+                                        />}
+                                    {certNames.length > 0 && certNames.map((option, index) => (
+                                        <FormSelectOption
+                                            key={index}
+                                            value={option}
+                                            label={option}
+                                        />
+                                    ))}
+                                </FormSelect>
+                            </div>
+                            <div title={_("Enter the full path on the server to and including the certificate file name")}>
+                                <Radio
+                                    id="certRadioFile"
+                                    className="ds-margin-top-lg"
+                                    label={_("Certificate File Location")}
+                                    name="certChoice"
+                                    isChecked={certRadioFile}
+                                    onChange={handleRadioChange}
+                                />
+                            </div>
+                            <div className={certRadioFile ? "ds-margin-top ds-radio-indent" : "ds-margin-top ds-radio-indent ds-disabled"}>
+                                <TextInput
+                                    type="text"
+                                    id="certFile"
+                                    aria-describedby="horizontal-form-name-helper"
+                                    name="certFile"
+                                    onChange={(e, value) => {
+                                        handleChange(e);
+                                    }}
+                                    validated={certRadioFile && certFile === "" ? ValidatedOptions.error : ValidatedOptions.default}
+                                />
+                            </div>
+                        </GridItem>
+                    </Grid>
+                </Form>
+            </Modal>
+        );
+    }
+}
+
+export class SecurityAddCACertModal extends React.Component {
+    render() {
+        const {
+            showModal,
+            closeHandler,
+            handleChange,
+            saveHandler,
+            spinning,
+            certName,
+            certFile,
+            certRadioFile,
+            certRadioSelect,
+            certRadioUpload,
+            handleRadioChange,
+            badCertText,
+            certNames,
+            certNicknames,
+            CACertNicknames,
+            // Select server cert
+            handleCertSelect,
+            selectCertName,
+            // File Upload
+            uploadValue,
+            uploadFileName,
+            uploadIsLoading,
+            uploadIsRejected,
+            handleFileInputChange,
+            handleTextOrDataChange,
+            handleFileReadStarted,
+            handleFileReadFinished,
+            handleClear,
+            handleFileRejected,
+        } = this.props;
+
+        let saveBtnName = _("Add Certificate");
+        const extraPrimaryProps = {};
+        if (spinning) {
+            saveBtnName = _("Adding Certificate ...");
+            extraPrimaryProps.spinnerAriaValueText = _("Saving");
         }
 
+        const certTextLabel = (
+            <div>
+                <Tooltip
+                    content={
+                        <div>
+                            {_("Paste the base64 encoded certificate that starts with \"-----BEGIN CERTIFICATE-----\" and ends with \"-----END CERTIFICATE-----\".  Make sure there are no special carriage return characters after each line.")}
+                        </div>
+                    }
+                >
+                    <div>{_("Upload PEM File, or Certificate Text")} <OutlinedQuestionCircleIcon /></div>
+                </Tooltip>
+            </div>
+        );
+
+
+        let title = _("Add Certificate Authority");
+        let desc = _("Add a CA Certificate to the security database.");
         let selectValidated = ValidatedOptions.default;
         if (certRadioSelect && certNames.length === 0) {
             selectValidated = ValidatedOptions.error;
@@ -201,7 +574,7 @@ export class SecurityAddCertModal extends React.Component {
                         key="confirm"
                         variant="primary"
                         onClick={() => {
-                            saveHandler(isCACert);
+                            saveHandler(true);
                         }}
                         isLoading={spinning}
                         spinnerAriaValueText={spinning ? _("Saving") : undefined}
@@ -209,7 +582,9 @@ export class SecurityAddCertModal extends React.Component {
                         isDisabled={
                             certName === "" || (certRadioFile && certFile === "") ||
                             (certRadioUpload && (uploadValue === "" || badCertText)) ||
-                            (certRadioSelect && certNames.length === 0)
+                            (certRadioSelect && certNames.length === 0) ||
+                            certNicknames.includes(certName) ||
+                            CACertNicknames.includes(certName)
                         }
                     >
                         {saveBtnName}
@@ -226,7 +601,7 @@ export class SecurityAddCertModal extends React.Component {
                         </Text>
                     </TextContent>
                     <Grid
-                        className="ds-margin-top-lg"
+                        className="ds-margin-top-sm"
                         title={_("Enter name/nickname of the certificate")}
                     >
                         <GridItem className="ds-label" span={3}>
@@ -241,11 +616,21 @@ export class SecurityAddCertModal extends React.Component {
                                 onChange={(e, str) => {
                                     handleChange(e);
                                 }}
-                                validated={certName === "" ? ValidatedOptions.error : ValidatedOptions.default}
+                                validated={
+                                    certName === "" ||
+                                    certNicknames.includes(certName) ||
+                                    CACertNicknames.includes(certName) ? ValidatedOptions.error : ValidatedOptions.default}
                             />
+                            {(certNicknames.includes(certName) || CACertNicknames.includes(certName)) && (
+                                <HelperText>
+                                    <HelperTextItem variant="error">
+                                        {_("Please use a unique certificate nickname")}
+                                    </HelperTextItem>
+                                </HelperText>
+                            )}
                         </GridItem>
                     </Grid>
-                    <Grid className="ds-margin-top">
+                    <Grid className="ds-margin-top-lg">
                         <GridItem span={12}>
                             <div title={_("Upload the contents of a PEM file from the client's system.")}>
                                 <Radio
@@ -298,7 +683,9 @@ export class SecurityAddCertModal extends React.Component {
                                 <FormSelect
                                     value={selectCertName}
                                     id="selectCertName"
-                                    onChange={handleCertSelect}
+                                    onChange={(e, str) => {
+                                        handleCertSelect(str);
+                                    }}
                                     aria-label="FormSelect Input"
                                     className="ds-cert-select"
                                     validated={selectValidated}
@@ -1103,6 +1490,22 @@ SecurityAddCertModal.defaultProps = {
     error: {},
 };
 
+SecurityAddCACertModal.propTypes = {
+    showModal: PropTypes.bool,
+    closeHandler: PropTypes.func,
+    handleChange: PropTypes.func,
+    saveHandler: PropTypes.func,
+    spinning: PropTypes.bool,
+    error: PropTypes.object,
+};
+
+SecurityAddCACertModal.defaultProps = {
+    showModal: false,
+    spinning: false,
+    saveHandler: () => {},
+    error: {},
+};
+
 SecurityAddCSRModal.propTypes = {
     showModal: PropTypes.bool,
     closeHandler: PropTypes.func,
diff --git a/src/cockpit/389-console/src/lib/security/securityTables.jsx b/src/cockpit/389-console/src/lib/security/securityTables.jsx
index fce4cb04e..a6c64e6c1 100644
--- a/src/cockpit/389-console/src/lib/security/securityTables.jsx
+++ b/src/cockpit/389-console/src/lib/security/securityTables.jsx
@@ -4,7 +4,6 @@ import {
     Grid,
     GridItem,
     Pagination,
-    PaginationVariant,
     SearchInput,
     Tooltip,
 } from '@patternfly/react-core';
@@ -94,10 +93,10 @@ class KeyTable extends React.Component {
     ];
 
     handleSort(_event, index, direction) {
-        const sortedRows = [...this.state.rows].sort((a, b) => 
+        const sortedRows = [...this.state.rows].sort((a, b) =>
             (a[index] < b[index] ? -1 : a[index] > b[index] ? 1 : 0)
         );
-        
+
         this.setState({
             sortBy: {
                 index,
@@ -127,7 +126,7 @@ class KeyTable extends React.Component {
                 >
                     <a className="ds-font-size-sm">{_("What is an orphan key?")}</a>
                 </Tooltip>
-                <Table 
+                <Table
                     className="ds-margin-top"
                     aria-label="orph key table"
                     variant="compact"
@@ -135,7 +134,7 @@ class KeyTable extends React.Component {
                     <Thead>
                         <Tr>
                             {columns.map((column, idx) => (
-                                <Th 
+                                <Th
                                     key={idx}
                                     sort={column.sortable ? {
                                         sortBy,
@@ -163,7 +162,7 @@ class KeyTable extends React.Component {
                                 )}
                                 {hasRows && (
                                     <Td isActionCell>
-                                        <ActionsColumn 
+                                        <ActionsColumn
                                             items={this.getActionsForRow(row)}
                                         />
                                     </Td>
@@ -224,7 +223,7 @@ class CSRTable extends React.Component {
     }
 
     handleSort(_event, index, direction) {
-        const sortedRows = [...this.state.rows].sort((a, b) => 
+        const sortedRows = [...this.state.rows].sort((a, b) =>
             (a[index] < b[index] ? -1 : a[index] > b[index] ? 1 : 0)
         );
         this.setState({
@@ -316,7 +315,7 @@ class CSRTable extends React.Component {
                         onClear={(evt) => this.handleSearchChange(evt, '')}
                     />
                 }
-                <Table 
+                <Table
                     className="ds-margin-top"
                     aria-label="csr table"
                     variant="compact"
@@ -324,7 +323,7 @@ class CSRTable extends React.Component {
                     <Thead>
                         <Tr>
                             {columns.map((column, idx) => (
-                                <Th 
+                                <Th
                                     key={idx}
                                     sort={column.sortable ? {
                                         sortBy,
@@ -352,7 +351,7 @@ class CSRTable extends React.Component {
                                 )}
                                 {hasRows && (
                                     <Td isActionCell>
-                                        <ActionsColumn 
+                                        <ActionsColumn
                                             items={this.getActionsForRow(row)}
                                         />
                                     </Td>
@@ -418,41 +417,11 @@ class CertTable extends React.Component {
     }
 
     handleSort(_event, columnIndex, direction) {
-        const sorted_rows = [];
-        const rows = [];
-        let count = 0;
-
-        // Convert the rows pairings into a sortable array
-        for (let idx = 0; idx < this.state.rows.length; idx += 2) {
-            sorted_rows.push({
-                expandedRow: this.state.rows[idx + 1],
-                1: this.state.rows[idx].cells[0].content,
-                2: this.state.rows[idx].cells[1].content,
-                3: this.state.rows[idx].cells[2].content,
-                issuer: this.state.rows[idx].issuer,
-                flags: this.state.rows[idx].flags
-            });
-        }
+        const rows = [...this.state.rows];
 
-        sorted_rows.sort((a, b) => (a[columnIndex + 1] > b[columnIndex + 1]) ? 1 : -1);
+        rows.sort((a, b) => (a.cells[columnIndex].content > b.cells[columnIndex].content) ? 1 : -1);
         if (direction !== SortByDirection.asc) {
-            sorted_rows.reverse();
-        }
-
-        for (const srow of sorted_rows) {
-            rows.push({
-                isOpen: false,
-                cells: [
-                    { content: srow[1] },
-                    { content: srow[2] },
-                    { content: srow[3] }
-                ],
-                issuer: srow.issuer,
-                flags: srow.flags,
-            });
-            srow.expandedRow.parent = count;
-            rows.push(srow.expandedRow);
-            count += 2;
+            rows.reverse();
         }
 
         this.setState({
@@ -534,18 +503,16 @@ class CertTable extends React.Component {
             rows.push(
                 {
                     isOpen: false,
-                    cells: [cert.attrs.nickname, cert.attrs.subject, cert.attrs.expires],
+                    cells: [
+                        { content: cert.attrs.nickname },
+                        { content: cert.attrs.subject },
+                        { content: cert.attrs.expires }
+                    ],
                     issuer: cert.attrs.issuer,
                     flags: cert.attrs.flags,
-
-                },
-                {
-                    parent: count,
-                    fullWidth: true,
-                    cells: [{ title: this.getExpandedRow(cert.attrs.issuer, cert.attrs.flags) }]
-                },
+                }
             );
-            count += 2;
+            count += 1;
         }
 
         this.setState({
@@ -587,7 +554,7 @@ class CertTable extends React.Component {
                         onChange={this.handleSearchChange}
                         onClear={(evt) => this.handleSearchChange(evt, '')}
                     />}
-                <Table 
+                <Table
                     aria-label="cert table"
                     variant='compact'
                 >
@@ -626,7 +593,7 @@ class CertTable extends React.Component {
                                         </Td>
                                     ))}
                                     <Td isActionCell>
-                                        <ActionsColumn 
+                                        <ActionsColumn
                                             items={this.getActionsForRow(row)}
                                         />
                                     </Td>
@@ -646,7 +613,7 @@ class CertTable extends React.Component {
                 </Table>
                 {hasRows &&
                     <Pagination
-                        itemCount={this.state.rows.length / 2}
+                        itemCount={this.state.rows.length}
                         widgetId="pagination-options-menu-bottom"
                         perPage={perPage}
                         page={page}
@@ -728,7 +695,7 @@ class CRLTable extends React.Component {
         if (direction !== SortByDirection.asc) {
             sorted_rows.reverse();
         }
-        
+
         for (const srow of sorted_rows) {
             rows.push({
                 isOpen: false,
@@ -806,14 +773,14 @@ class CRLTable extends React.Component {
                     onChange={this.handleSearchChange}
                     onClear={(evt) => this.handleSearchChange(evt, '')}
                 />
-                <Table 
+                <Table
                     aria-label="CRL Table"
                     variant="compact"
                 >
                     <Thead>
                         <Tr>
                             {this.state.columns.map((column, idx) => (
-                                <Th 
+                                <Th
                                     key={idx}
                                     sort={column.sortable ? {
                                         sortBy: this.state.sortBy,
@@ -836,7 +803,7 @@ class CRLTable extends React.Component {
                                     ))}
                                     {this.state.hasRows && (
                                         <Td isActionCell>
-                                            <ActionsColumn 
+                                            <ActionsColumn
                                                 items={this.getActionsForRow(row)}
                                             />
                                         </Td>
diff --git a/src/cockpit/389-console/src/security.jsx b/src/cockpit/389-console/src/security.jsx
index c4504066b..3aead0449 100644
--- a/src/cockpit/389-console/src/security.jsx
+++ b/src/cockpit/389-console/src/security.jsx
@@ -60,6 +60,8 @@ export class Security extends React.Component {
             primaryCertName: '',
             serverCertNames: [],
             serverCerts: [],
+            CACerts: [],
+            CACertNames: [],
             isMinSSLOpen: false,
             isMaxSSLOpen: false,
             isClientAuthOpen: false,
@@ -330,9 +332,14 @@ export class Security extends React.Component {
                 .spawn(cmd, { superuser: true, err: "message" })
                 .done(content => {
                     const certs = JSON.parse(content);
+                    const certNames = [];
+                    for (const cert of certs) {
+                        certNames.push(cert.attrs.nickname);
+                    }
                     this.setState(() => (
                         {
                             CACerts: certs,
+                            CACertNames: certNames,
                         }), this.loadCSRs
                     );
                 })
@@ -1226,6 +1233,9 @@ export class Security extends React.Component {
                                     ServerKeys={this.state.serverOrphanKeys}
                                     addNotification={this.props.addNotification}
                                     certDir={this.props.certDir}
+                                    certNicknames={this.state.serverCertNames}
+                                    CACertNicknames={this.state.CACertNames}
+                                    reloadCerts={this.loadCerts}
                                 />
                             </Tab>
                             <Tab eventKey={2} title={<TabTitleText>{_("Cipher Preferences")}</TabTitleText>}>
diff --git a/src/lib389/lib389/cli_conf/security.py b/src/lib389/lib389/cli_conf/security.py
index 18e444c4c..bc3a6ddaf 100644
--- a/src/lib389/lib389/cli_conf/security.py
+++ b/src/lib389/lib389/cli_conf/security.py
@@ -86,9 +86,9 @@ def _security_generic_get(inst, basedn, log, args, attrs_map):
             val = ""
         result[props.attr.lower()] = val
     if args.json:
-        print(json.dumps({'type': 'list', 'items': result}, indent=4))
+        log.info(json.dumps({'type': 'list', 'items': result}, indent=4))
     else:
-        print('\n'.join([f'{attr}: {value or ""}' for attr, value in result.items()]))
+        log.info('\n'.join([f'{attr}: {value or ""}' for attr, value in result.items()]))
 
 
 def _security_generic_set(inst, basedn, log, args, attrs_map):
@@ -228,10 +228,10 @@ def security_ciphers_set(inst, basedn, log, args):
 def security_ciphers_get(inst, basedn, log, args):
     enc = Encryption(inst)
     if args.json:
-        print({'type': 'list', 'items': enc.ciphers})
+        log.info({'type': 'list', 'items': enc.ciphers})
     else:
         val = ','.join(enc.ciphers)
-        print(val if val != '' else '<undefined>')
+        log.info(val if val != '' else '<undefined>')
 
 
 def security_ciphers_list(inst, basedn, log, args):
@@ -247,12 +247,13 @@ def security_ciphers_list(inst, basedn, log, args):
         lst = enc.ciphers
 
     if args.json:
-        print(json.dumps({'type': 'list', 'items': lst}, indent=4))
+        log.info(json.dumps({'type': 'list', 'items': lst}, indent=4))
     else:
         if lst == []:
             log.getChild('security').warn('List of ciphers is empty')
         else:
-            print(*lst, sep='\n')
+            for item in lst:
+                log.info(item)
 
 
 def security_disable_plaintext_port(inst, basedn, log, args, warn=True):
@@ -312,7 +313,10 @@ def cert_list(inst, basedn, log, args):
     certmgr = CertManager(instance=inst)
     certs = certmgr.list_certs()
     if not certs:
-        log.info("No certificates found.")
+        if args.json:
+            log.info(json.dumps([], indent=4))
+        else:
+            log.info("No certificates found.")
         return
 
     if args.json:
@@ -328,7 +332,10 @@ def cacert_list(inst, basedn, log, args):
     certmgr = CertManager(instance=inst)
     ca_certs = certmgr.list_ca_certs()
     if not ca_certs:
-        log.info("No CA certificates found.")
+        if args.json:
+            log.info(json.dumps([], indent=4))
+        else:
+            log.info("No CA certificates found.")
         return
 
     if args.json:
-- 
2.53.0

